### PR TITLE
Support script_fields query

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

Allow support of Elasticsearch [script fields](https://www.elastic.co/guide/en/elasticsearch/reference/7.5/search-request-body.html#request-body-search-script-fields) in search body.

Fix https://github.com/kuzzleio/kuzzle/issues/1541

### How should this be manually tested?

Use this script, you should see the `fields` property on each hit:
```
const {
  Kuzzle,
  WebSocket
} = require('kuzzle-sdk')

const kuzzle = new Kuzzle(new WebSocket('localhost'))

kuzzle.on('networkError', err => console.log(err));

(async () => {
  try {
    await kuzzle.connect();
    const query = {
      "script_fields": {
        "ageValue": {
          "script": {
            "lang": "painless",
            "source": "doc['age'].value"
          }
        }
      },
      "query": {}
    };

    if (! await kuzzle.collection.exists('index', 'collection')) {
      await kuzzle.index.create('index');
      await kuzzle.collection.create('index', 'collection');
      await kuzzle.document.create('index', 'collection', { age: 21 });
      await kuzzle.document.create('index', 'collection', { age: 42 });
      await kuzzle.collection.refresh('index', 'collection');
    }

    const res = await kuzzle.document.search('index', 'collection', query);
    console.log(JSON.stringify(res.hits, null, 2));
  }
  catch (error) {
    console.log(error);
  }
  finally {
    kuzzle.disconnect();
  }
})();
```

### Boyscout

 - async/await and Array.map

### CODE REVIEW

File: lib/services/elasticsearch.js [hunk 1]
```diff
@@ -222,41 +222,37 @@ class ElasticSearch extends Service {
...
-            key = scrollCachePrefix + this._kuzzle.constructor.hash(
-              body._scroll_id
-            );
-
-          return this._kuzzle.cacheEngine.internal.psetex(key, ttl, 0)
-            .then(() => this._formatSearchResult(body));
-        }
+    let body;
+    try {
+      ({ body } = await this._client.search(esRequest));
```
Comment (kuzzle) [inline lib/services/elasticsearch.js]: MAJOR Extract the assignment of "{ body }" from this expression. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AAssignmentWithinCondition)

**Comment by kuzzle** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1678](https://github.com/kuzzleio/kuzzle/blob/c1410f9c4c3b0f2101dc226e118421dcb916278c/lib/services/elasticsearch.js#L1678): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1732](https://github.com/kuzzleio/kuzzle/blob/c1410f9c4c3b0f2101dc226e118421dcb916278c/lib/services/elasticsearch.js#L1732): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1812](https://github.com/kuzzleio/kuzzle/blob/c1410f9c4c3b0f2101dc226e118421dcb916278c/lib/services/elasticsearch.js#L1812): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: lib/services/elasticsearch.js [hunk 2]
```diff
@@ -222,41 +222,37 @@ class ElasticSearch extends Service {
...
-            key = scrollCachePrefix + this._kuzzle.constructor.hash(
-              body._scroll_id
-            );
-
-          return this._kuzzle.cacheEngine.internal.psetex(key, ttl, 0)
-            .then(() => this._formatSearchResult(body));
-        }
+    let body;
+    try {
+      ({ body } = await this._client.search(esRequest)); // NOSONAR
```
Comment (Aschen) [inline lib/services/elasticsearch.js]: It's annoying to have to add this comment because Sonarqube does not like this syntax.

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [elasticsearch.js#L1678](https://github.com/kuzzleio/kuzzle/blob/4d751a4cdcfb2d74f7874875f25bc018c3658f9c/lib/services/elasticsearch.js#L1678): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1732](https://github.com/kuzzleio/kuzzle/blob/4d751a4cdcfb2d74f7874875f25bc018c3658f9c/lib/services/elasticsearch.js#L1732): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [elasticsearch.js#L1812](https://github.com/kuzzleio/kuzzle/blob/4d751a4cdcfb2d74f7874875f25bc018c3658f9c/lib/services/elasticsearch.js#L1812): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/services/elasticsearch.js (hunk 2 above)
Comment (alexandrebouthinon) [reply lib/services/elasticsearch.js]: Did you try:
```js
       { body } = (await this._client.search(esRequest)); // NOSONAR 
```
Maybe this one is allowed by Sonar :slightly_smiling_face:

Comment (Aschen) [reply lib/services/elasticsearch.js]: This syntax is not valid javascript unfortunately :(

Comment (scottinet) [reply lib/services/elasticsearch.js]: I already notified Florient to ask if he could update our sonarqube servers, it's stuck to a very old version.
In the meantime... yeah, more and more modern JS gets rejected by sonar

**Comment by Aschen** [comment]
> @scottinet 
Currently the script are executed but the result is never sent back to the client.  
I agree that this could be a dangerous feature and at least administrators should be able to control who has access to it.

But I also agree with @alexandrebouthinon that we should try to abstract ourselves from ES so I didn't think this feature should be available through our API at all because it will involve a lot of specific code. 
IMHO, developers should use the [ESClient constructor](https://docs.kuzzle.io/core/2/plugins/plugin-context/constructors/esclient/) inside plugins if they want to use scripts.

**Comment by alexandrebouthinon** [comment]
@Aschen @scottinet Should we close this PR?

**Comment by Aschen** [comment]
> @Aschen @scottinet Should we close this PR?

I think we can close this PR but we have to open another to prevent usage of `script_fields`

**Comment by cklinx** [comment]
> 
> 
> > @Aschen @scottinet Should we close this PR?
> 
> I think we can close this PR but we have to open another to prevent usage of `script_fields`

Sorry, but before close it i think should need some other way to enable it if someone needs. There are some variable that could be get or calculated only with script.

**Comment by Aschen** [comment]
Usage of scripts was feasible but undocumented. We decided to block them in v2 because it can lead to DoS.
See https://github.com/kuzzleio/kuzzle/issues/1541#issuecomment-563921657

**Pull request closed without merging**